src/server/enrichment/validation/index.js

```diff
@@ -5,100 +5,144 @@ const { validTargetDb } = require('./validity-info');
-    formData.target = convertGConvertNames(initialTarget);
-    const invalidInfo = { invalidTargetDb: undefined, invalidOrganism: undefined };
-    const queryVal = formData.query;
-    if (!Array.isArray(queryVal)) {
-      reject(new Error('ERROR: genes should be an array'));
-    }
-    formData.query = queryVal.join(" ");
-    if (!validOrganism.includes(formData.organism)) {
-      invalidInfo.invalidOrganism = formData.organism;
-    }
-    if (!validTargetDb.includes(formData.target)) {
-      invalidInfo.invalidTargetDb = formData.target;
-    }
-    if (invalidInfo.invalidOrganism != undefined || invalidInfo.invalidTargetDb != undefined) {
-      reject(new InvalidInfoError(invalidInfo.invalidOrganism, invalidInfo.invalidTargetDb, ''));
-    }
+  const form = _.assign( {},
+    defaultOptions,
+    JSON.parse( JSON.stringify( userOptions ) ),
+    { query: query }
+  );
+
+  if (!Array.isArray( form.query )) {
+    throw new InvalidParameterError( 'Invalid genes: Must be an array' );
+  }
+  if ( !validOrganism.includes( form.organism.toLowerCase() ) ) {
+    throw new InvalidParameterError( 'Invalid organism' );
+  }
+  if ( !validTargetDb.includes( form.target.toUpperCase() ) ) {
+    throw new InvalidParameterError( 'Invalid target' );
+  }
+
+  form.target = mapDBNames( form.target );
+  form.query = form.query.join(" ");
 
-    fetch(gConvertURL, {
-      method: 'post',
-      body: qs.stringify(formData)
-    }).then(gConvertResponse => gConvertResponse.text())
-      .then(body => {
-        const geneInfoList = _.map(body.split('\n'), ele => { return ele.split('\t'); });
-        geneInfoList.splice(-1, 1); // remove last element ''
-        const unrecognized = new Set();
-        let duplicate = {};
-        const previous = new Map();
-        let geneInfo = new Set();
-        const initialAliasIndex = 1;
-        const convertedAliasIndex = 3;
-        _.forEach(geneInfoList, info => {
-          const convertedAlias = info[convertedAliasIndex];
-          let initialAlias = info[initialAliasIndex];
-          initialAlias = cleanUpEntrez(initialAlias);
-          if (convertedAlias === 'N/A') {
-            unrecognized.add(initialAlias);
-          } else {
-            if (!previous.has(convertedAlias)) {
-              previous.set(convertedAlias, initialAlias);
-            } else {
-              if (!(convertedAlias in duplicate)) {
-                duplicate[convertedAlias] = new Set([previous.get(convertedAlias)]);
-              }
-              duplicate[convertedAlias].add(initialAlias);
-            }
-            geneInfo.add(JSON.stringify({initialAlias: initialAlias, convertedAlias: convertedAlias}));
-          }
-        });
-        for (const initialAlias in duplicate) {
-          duplicate[initialAlias] = Array.from(duplicate[initialAlias]);
+  return form;
+};
+
+const bodyHandler = body =>  {
+  const geneInfoList = _.map(body.split('\n'), ele => { return ele.split('\t'); });
```

Comment:
This is better than before, but maybe we could have a student do some refactoring on this function.  It could be cleaned up and optimised a bit more

Reply:
Sometime later -- outside of this pr